jormungandr/src/settings/start/trusted_peer.rs

```diff
@@ -0,0 +1,146 @@
+use crate::{network::p2p::Id, settings::start::config::Address};
+use multiaddr::AddrComponent;
+use serde::{de::Visitor, Deserialize, Deserializer, Serialize, Serializer};
+use std::{
+    fmt,
+    iter::FromIterator,
+    net::{SocketAddr, ToSocketAddrs},
+};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct TrustedPeer {
+    pub address: TrustedAddress,
+    pub id: Id,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TrustedAddress(pub multiaddr::Multiaddr);
+
+custom_error! {pub AddressError
+    DnsLookupError { source: std::io::Error } = "failed to resolve DNS name {source}",
+    NoPortSpecified = "no TCP port specified",
+    NoAppropriateDNSFound = "the address was resolved, but it doesn't provide IPv4 or IPv6 addresses",
+    UnsupportedProtocol = "the provided protocol is unsupported, please use one of ip4/ip6/dns4/dns6",
+}
+
+impl TrustedAddress {
+    pub fn to_addresses(&self) -> Result<Vec<Address>, AddressError> {
+        let mut components = self.0.iter();
+        let protocol = components.next();
+
+        if let Some(AddrComponent::IP4(_)) | Some(AddrComponent::IP6(_)) = protocol {
+            return Ok(vec![Address(
+                poldercast::Address::new(self.0.clone()).unwrap(),
+            )]);
+        }
+
+        let port = match components.next() {
+            Some(AddrComponent::TCP(port)) => port,
+            _ => return Err(AddressError::NoPortSpecified),
+        };
+
+        let addresses: Vec<AddrComponent> = match protocol {
+            Some(AddrComponent::DNS4(fqdn)) => format!("{}:{}", fqdn, port)
+                .to_socket_addrs()
+                .map_err(|e| AddressError::DnsLookupError { source: e })?
+                .into_iter()
+                .filter_map(|r| match r {
+                    SocketAddr::V4(addr) => Some(AddrComponent::IP4(*addr.ip())),
+                    _ => None,
+                })
+                .collect(),
+            Some(AddrComponent::DNS6(fqdn)) => format!("{}:{}", fqdn, port)
```

Comment:
It's... weird that multiaddr does not support just `dns/` which would resolve both A and AAAA records. Which seems to be what's going to happen under the hood anyway, except that you discard the addresses of the wrong type.

Reply:
Converting from multiaddr to an FQDN URL for `hyper`, as per https://github.com/input-output-hk/jormungandr/pull/1164#discussion_r349581675, would erase the IP address preference, so I'm not sure what would be more proper here.

Reply:
Theoretically, `HttpConnector::set_local_address` [in hyper](https://docs.rs/hyper/0.12.35/hyper/client/connect/struct.HttpConnector.html#method.set_local_address) can be used to restrict connections to IPv4 or IPv6, but I'm not sure it works with the internal resolver.

Reply:
Either approach requires us to make the whole system treat FQDNs the same way as IPs and probably bring everything to `network-grpc` (see https://github.com/input-output-hk/jormungandr/pull/1164#discussion_r349745898).